src/app/shared/spaces.service.ts

```diff
@@ -1,72 +1,84 @@
 export class SpacesService implements Spaces {
 
   static readonly RECENT_SPACE_LENGTH = 8;
-  private static readonly RECENT_SPACE_KEY = 'recentSpaceIds';
 
-  addRecent: Subject<Space> = new Subject<Space>();
   private _current: Observable<Space>;
-  private _recent: ConnectableObservable<Space[]>;
+  private _recent: ReplaySubject<Space[]> = new ReplaySubject<Space[]>(1);
 
   constructor(
     private contexts: Contexts,
-    private localStorage: LocalStorageService,
     private spaceService: SpaceService,
     private broadcaster: Broadcaster,
     private profileService: ProfileService
   ) {
+    this.initCurrent();
+    this.initRecent();
+  }
+
+  private initCurrent(): void {
     this._current = Observable.merge(
-      contexts.current
-        .map(val => val.space),
-      this.broadcaster.on<Space>('spaceUpdated'));
-    // Create the recent context list
+      this.contexts.current.map(val => val.space),
+      this.broadcaster.on<Space>('spaceUpdated')
+    );
+  }
+
+  private initRecent(): void {
+    // load existing recent spaces from profile.store.recentSpaces
+    this.loadRecent().subscribe(spaces => {
+      this._recent.next(spaces);
+    });
 
-    this._recent = Observable.merge(
-      this.broadcaster.on<Space>('spaceChanged'),
-      this.addRecent)
-      // Map from the context being added to an array of recent contexts
-      // The scan operator allows us to access the list of recent contexts and add ours
-      .scan((recent, s) => {
-        // First, check if this context is already in the list
-        // If it is, remove it, so we don't get duplicates
-        for (let i = recent.length - 1; i >= 0; i--) {
-          if (recent[i].id === s.id) {
-            recent.splice(i, 1);
+    // update _recent with newly seen space when spaceChanged is emitted
+    this.broadcaster.on<Space>('spaceChanged')
+      .flatMap((changedSpace: Space) => {
+        return this._recent.first().map((recentSpaces: Space[]) => {
+          let index: number = this.findRecentIndexById(recentSpaces, changedSpace.id);
+          // if changedSpace exists in _recent
+          if (index !== -1) {
+            // move the space to the front of the list if it is already in _recent
+            if (recentSpaces[0].id !== changedSpace.id) {
+              recentSpaces.splice(index, 1);
+              recentSpaces.unshift(changedSpace);
+              this.saveRecent(recentSpaces);
+              this._recent.next(recentSpaces);
+            }
+            // otherwise, no operation required; changedSpace is already index 0
+          // if changedSpace is new to the recent spaces
+          } else {
+            recentSpaces.unshift(changedSpace);
+            // trim _recent if required to ensure it is length =< 8
+            if (recentSpaces.length > SpacesService.RECENT_SPACE_LENGTH) {
+              recentSpaces.pop();
+            }
+            this.saveRecent(recentSpaces);
+            this._recent.next(recentSpaces);
```

Comment:
It looks like this is called after every `saveRecent(...)`, maybe it can be combined as one function?

Reply:
Sure, I could put it into `saveRecent`